### PR TITLE
Add trade market sell amount display/overlay

### PR DESCRIPTION
**Pull request by byBackfish**
Pretty self explanatory
image

### CODE REVIEW

**Comment by magicus** [comment]
Is the idea to help the user set the right amount?

Can't you take it a step further, so if you e.g. right click on the amount button, the determined amount is sent as a chat message?

**Comment by byBackfish** [comment]
> Is the idea to help the user set the right amount?
> 
> Can't you take it a step further, so if you e.g. right click on the amount button, the determined amount is sent as a chat message?

Good idea. I do not know why i did not think of this earlier. Ty for the suggestion

**Comment by ryanzhoudev** [comment]
https://feedback.wynntils.com/posts/21/a-sell-all-button

File: src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class TradeMarketOverlay implements Listener {
+
+    private final Pattern ITEM_NAME_PATTERN = Pattern.compile(".*Selling §f\\d+ (.*) for .* Each");
+
+    @SubscribeEvent
+    public void onRender(GuiOverlapEvent.ChestOverlap.HandleMouseClick e) {
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java]: Can we get some new lines? This code feels so crammed.

File: src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        if (McIf.player() == null) return;
+        if (McIf.player().openContainer == null) return;
+        InventoryBasic inventory = (InventoryBasic) e.getGui().getLowerInv();
+        if (!inventory.getName().contains("What would you like to sell?")) return;
+        if(e.getSlotId() != 11) return;
+        ItemStack itemStack = inventory.getStackInSlot(10);
+        if (itemStack == ItemStack.EMPTY) return;
+        if (itemStack.getDisplayName().contains("Click an Item to Sell")) return;
+        System.out.println(itemStack.getDisplayName());
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java]: Debug line?

**Review by kristofbolyai — CHANGES_REQUESTED** [review]

**Comment by magicus** [comment]
> https://feedback.wynntils.com/posts/21/a-sell-all-button

@DonkeyBlaster Thanks for doing the work with confirming approval from the Wynncraft team, for this and other potential additions!

**Comment by magicus** [comment]
@byBackfish You misunderstood me. I meant that Wynntils should take care of sending the reply in chat for the user. That is, if the user right clicks on the plus, Wynntils will do the following:
 * Fake a click on the "+" icon.
 * Send the amount of items as a chat message sent *from* the user.

I realize right-clicking might be both harder to do, and less visible, so maybe a better way to trigger this would be to add an additional "+" icon below (or above, whichever looks best), give it the tooltip "Add all items of this kind", and when that is clicked, do the "dance" above.

If this needs to much knowledge of Wynntils inner workings for you to pull off, you can go with the current solution. But if possible, I think it would be much nicer to have.

**Comment by magicus** [comment]
(Also, that new "+" item could have the number you are going to add as index/"amount") so you can directly see the result of your action before doing it.

**Comment by byBackfish** [comment]
@magicus oh alright. I tried my best, hope this is fine
image

**Comment by byBackfish** [comment]
Actually, I thought an Emerald would look nicer than the "+", but how about making it a green checkmark?

**Comment by magicus** [comment]
Yeah, the "+" was just a suggestion. If there were a kind of "++" icon, I think that'd be the best, but whatever. I think the check mark should be reserved for executing the sale. 

The drawback with the emerald is that it is already used to signal setting the price. 

Maybe using a full pouch icon?

File: src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    }
+
+    public NBTTagCompound buildNBT(InventoryBasic inventory) {
+        String itemName = getItemName(inventory);
+
+        NBTTagCompound nbt = new NBTTagCompound();
+        NBTTagCompound tag = new NBTTagCompound();
+
+        nbt.setString("id", "minecraft:emerald");
+        nbt.setInteger("Count", 1);
```
Comment (magicus) [inline src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java]: Try setting this to `getAmount(itemName)`. It should work up to 64 at least, I believe. I don't know if our code to display higher values is general enough to pick this up, but maybe. It's better than nothing, I think.

File: src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+        NBTTagCompound display = new NBTTagCompound();
+        display.setString("Name", "§aSell All");
+
+        NBTTagList lore = new NBTTagList();
+        lore.appendTag(new NBTTagString("§7Click to sell all items in your inventory"));
```
Comment (magicus) [inline src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java:79]:
```suggestion
        lore.appendTag(new NBTTagString("§7Click to sell all " + getAmount(itemName) + " items in your inventory"));
```

**Comment by magicus** [comment]
If you want to test the ingredient pouch idea, set `id` to `minecraft:diamond_axe` and `Damage` to `93`. Unfortunately, I did not have a character with a full ingredient pouch (and no time to fill one up right now), so I can't give you a damage value for the full pouch icon.

**Comment by byBackfish** [comment]
Yea, I have those damage values but thanks! Wouldn't it be better to make it a full emerald pouch instead of a full ingredient pouch?

**Comment by magicus** [comment]
Maaaaaybe...? But I think my point here is that the first two slots is what you sell (item + amount), and the third is the price. And that "price" is associated with emeralds, but the other two are not. So I think the association with emeralds is a bit odd; it feels more like "set a maximum price" or so, rather than "sell all that I got". 

Ideally, I think we should have some kind of stacking +:es, something along these lines: http://svn.slimdevices.com/repos/jive/7.7/trunk/squeezeplay/assets/MasterIcons/icon_add_all.png. But creating new icons is tricky and kind of out of the scope for this PR.

**Comment by byBackfish** [comment]
Once the count is above 64, the item just simply does not display. It now uses 64 if the amount would be above 64.

image

**Comment by magicus** [comment]
I think this looks great! :-)

**Review by magicus — DISMISSED** [review]

**Comment by magicus** [comment]
(An alternative would be to set count to 1 if above 64, to have no number at all instead of an incorrect. But I'm not even sure that's better.)

**Review by ryanzhoudev — DISMISSED** [review]
Very nice

Actually, you could make a few of the methods in TradeMarketOverlay private. (getAmount, getItemName, buildNBT)

File: src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+    @SubscribeEvent
+    public void onGuiUpdate(GuiOverlapEvent.ChestOverlap.DrawScreen e) {
+        InventoryBasic inventory = (InventoryBasic) e.getGui().getLowerInv();
+        if (!inventory.getName().startsWith("What would you like to sell?")) {
+            shouldSend = false;
+            return;
+        }
+        if (inventory.getStackInSlot(20) != ItemStack.EMPTY && !inventory.getStackInSlot(20).getDisplayName().isEmpty()) return;
+
```
Comment (magicus) [inline src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java:32]: Thinking about it a bit more... Maybe it only makes sense to display this icon if you have more than one item to sell? So if `getAmount(itemName) <= 1` you'd just return here. What do you think? Or do we always want this icon, even if you have only a single object of that kind?

**Review by ryanzhoudev — DISMISSED** [review]

**Review by magicus — DISMISSED** [review]

File: src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    @SubscribeEvent
+    public void onClick(GuiOverlapEvent.ChestOverlap.HandleMouseClick e) {
+        InventoryBasic inventory = (InventoryBasic) e.getGui().getLowerInv();
+        if (!inventory.getName().contains("What would you like to sell?")) return;
+        if (e.getSlotId() != 20) return;
+
+        String itemName = getItemName(inventory);
+        if(itemName == null) return;
```
Comment (kristofbolyai) [inline src/main/java/com/wynntils/modules/utilities/overlays/inventories/TradeMarketOverlay.java]:
```suggestion
        if (itemName == null) return;
```

**Review by kristofbolyai — APPROVED** [review]

**Review by magicus — APPROVED** [review]

**Pull request merged**